src/util/auth.ts

```diff
@@ -0,0 +1,92 @@
+        // (saves time if the route doesn't have route params)
+        if (routes.includes(uri.toLowerCase()))
+            return { route: uri.toLowerCase(), id: "" };
+        else {
+            // split uri and remove 1st empty position
+            const uriSlice = uri.split("/");
+            uriSlice.shift();
+            // extract id (can be any resource id but only user
+            // ids will be used in case of a user-protected route)
+            const id = uriSlice.length > 1 ? uriSlice[1] : "";
+
+            // lowercase all array elements
+            for (let i = 0; i < uriSlice.length; i++)
+                uriSlice[i] = uriSlice[i].toLowerCase();
+
+            // get base path
+            const basePath = uriSlice[0];
+
+            // if basepath doesn't exist, then route doesn't exist
+            if (!basePaths.includes(basePath)) return { route: "", id: "" };
+            else {
+                // remove the id variable to something consistent
+                if (id) uriSlice[1] = "{id}";
+                return {
+                    route: "/" + uriSlice.join("/"),
+                    id: id ? id : ""
+                };
+            }
+        }
+    },
+
+    authorize: (
+        route: string,
+        method: string,
+        role: string,
+        requestUserId: string,
+        resourceUserId: string
+    ): boolean => {
+        if (authConfig[route] && authConfig[route][method]) {
+            const routeConfig = authConfig[route][method];
+
+            switch (routeConfig.scope) {
+                case "INTERNAL":
+                    // reject right away since internal requests won't pass by the
+                    // introspection route
+                    return false;
+                case "PUBLIC":
+                    // accept right away since public requests won't pass by the
+                    // introspection route
+                    return true;
+                case "ADMIN":
+                    // check if role is admin
+                    return role === "ADMIN";
+                case "USER":
+                    // check if role is user or higher
+                    const authorizedInUserScope =
+                        role === "ADMIN" || role === "USER";
+
+                    if (routeConfig.userProtected) {
+                        // verify that the correct user is modifying the resource
+                        if (
+                            resourceUserId &&
+                            requestUserId === resourceUserId &&
+                            authorizedInUserScope
+                        )
+                            return true;
+                        else return false;
+                    }
+
+                    return authorizedInUserScope;
+                default:
+                    return false;
+            }
+        }
+
+        // default to rejection
+        // i.e. all unregistered routes are internal by default
+        return false;
```

Comment:
can't we flip this:
instead of having the big block of code in an `if(x)` statement, and then returning false as an alternative,
we can have a check at the top of the method for `if(!x)` and then have the block of code not be in an if statement.

Let me know if this makes sense.

Reply:
not sure what difference that makes

Reply:
Cleaner to read and understand. doesn't make a functional difference